### PR TITLE
Task edit about me

### PR DESCRIPTION
**Pull request by rrrichard**
let me out of this story

### CODE REVIEW

File: php/admin_edit.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
```
Comment (katestarks) [inline php/admin_edit.php]: Delete this file

File: php/admin_page.php [hunk 2]
```diff
@@ -1,4 +1,24 @@
...
+require_once '../db/db_query.php';
+
+
+$db = getDbConnection();
+
+$newParagraph = $_POST['edit_form'];
+$paragraphs = addAboutMe($db);
+$editDropdownResults = editParagraphDropdown($paragraphs);
+
+if ($_POST['submit'] == 'Edit'){
```
Comment (katestarks) [inline php/admin_page.php]: Use (isset())

File: php/admin_page.php [hunk 3]
```diff
@@ -1,4 +1,24 @@
...
+    $editChoice = $_POST['editSelect'];
+    $getEdit = getEdit($db, $editChoice);
+    $pasteEdit = pasteEdit($getEdit);
+}
+if ($_POST['submit'] == 'Submit'){
```
Comment (katestarks) [inline php/admin_page.php]: Use (isset())

File: php/admin_page.php [hunk 4]
```diff
@@ -28,17 +48,24 @@
         </div>
         <div class="formSize">
             <p>Edit paragraph</p>
-            <form>
-                <textarea rows="6" cols="80" name="edit_form"></textarea>
```
Comment (katestarks) [inline php/admin_page.php:32]: Don't need rows and columns, these should be set in css

File: php/admin_page.php [hunk 5]
```diff
@@ -28,17 +48,24 @@
...
+                <div class="edit_button">
+                    <input type="submit" name="submit" value="Edit">
+                </div>
+            </form>
+
+            <form method="post" class="formy editParaRight" action="admin_page.php">
+                <textarea class="edit_submit" rows="6" cols="80" name="edit_form"><?php echo $pasteEdit; ?></textarea>
+                <?php if (isset($editChoice)){
+                    echo '<input type="hidden" name="editId" value=' . $editChoice . ' />';
+                }?>
```
Comment (katestarks) [inline php/admin_page.php]: put the php close tag on a new line for readability

File: php/functions.php [hunk 6]
```diff
@@ -52,22 +47,76 @@ function addParagraphToDb (PDO $db, string $addSubmit) : void{
...
+ * @param PDO $db calls the database from the db_query because this function requires it
+ *
+ * @param $editChoice string outputs a number of the id that will be used by the WHERE in MySQL
+ *
+ * @param $newParagraph string is a variable that gets the text that was submitted in the 'edit paragraph' text area
+ *
+ * @return bool the function will only be called when the user press 'submit' and if true, will update the selected paragraph with the new paragraph
+ */
+function editParagraph(PDO $db, string $editChoice, string $newParagraph) : bool {
+    $query = $db->prepare("UPDATE `about_me` SET `paragraph`= :newParagraph WHERE `id`= :oldParagraphId;");
```
Comment (katestarks) [inline php/functions.php]: Is 'old paragraph id' semantic since the id doesn't change when you update the paragraph?

File: php/admin_page.php [hunk 7]
```diff
@@ -1,4 +1,24 @@
 <?php
+require_once 'functions.php';
+require_once '../db/db_query.php';
+
+
```
Comment (katestarks) [inline php/admin_page.php]: Delete extra line

File: php/admin_page.php [hunk 8]
```diff
@@ -1,4 +1,24 @@
...
+$db = getDbConnection();
+
```
Comment (katestarks) [inline php/admin_page.php]: Delete extra line

File: php/admin_page.php [hunk 9]
```diff
@@ -1,4 +1,24 @@
...
+if ($_POST['submit'] == 'Edit'){
+    $editChoice = $_POST['editSelect'];
+    $getEdit = getEdit($db, $editChoice);
+    $pasteEdit = pasteEdit($getEdit);
+}
+if ($_POST['submit'] == 'Submit'){
+    $editChoice = $_POST['editId'];
+    editParagraph($db, $editChoice, $newParagraph);
+}
+
```
Comment (katestarks) [inline php/admin_page.php:42]: Delete extra line

File: php/admin_page.php [hunk 10]
```diff
@@ -28,17 +48,24 @@
...
-                    <option>paragraph 2</option>
-                    <option>paragraph 3</option>
+                    <option selected="selected" value="choose">Choose Paragraph</option>
+                    <?php echo $editDropdownResults; ?>
                 </select>
+                <div class="edit_button">
+                    <input type="submit" name="submit" value="Edit">
+                </div>
+            </form>
+
```
Comment (katestarks) [inline php/admin_page.php]: Delete extra line

File: php/tests.php [hunk 11]
```diff
@@ -41,4 +41,58 @@ public function testAddParagraphsMalformed(){
...
+        $this->assertEquals($expected, $case);
+    }
+
+    public function testEditParagraphDropdownMalformed(){
+        $input = 3;
+        $this->expectException(TypeError::class);
+        editParagraphDropdown($input);
+    }
+
+    public function testPasteEditSuccess(){
```
Comment (katestarks) [inline php/tests.php]: It looks like your function does the same thing whether the case is a success or failure which doesn't seem right?

File: php/functions.php [hunk 12]
```diff
@@ -52,22 +47,76 @@ function addParagraphToDb (PDO $db, string $addSubmit) : void{
...
+/**
+ * this function enables the html page to call and put the paragraph that is chosen from the dropdown and into the textarea for edit
+ *
+ * @param array $getEdit is the variable that retrieves the output of the getEdit function, which is the selected paragraph in the dropdown
+ *
+ * @return string returns a variable that can be assigned in the admin page and can be put into the textarea, the textarea will be populated by the chosen paragraph
+ */
+function pasteEdit(array $getEdit) : string {
+    $editPopulate = '';
+    if (is_string($getEdit['paragraph']) &&array_key_exists('paragraph', $getEdit)) {
```
Comment (katestarks) [inline php/functions.php]: don't need to check whether a key is a string?

**Review by katestarks — CHANGES_REQUESTED** [review]
Currently if I just press submit before pre-populating the text area it throws an error - should address this one

Your CMS allows whitespaces before and after a paragraph; this means it also allows blank text in a paragraph

File: php/functions.php (hunk 6 above)
Comment (rrrichard) [reply php/functions.php]: not semantic, changed the name

File: php/admin_page.php [hunk 13]
```diff
@@ -1,4 +1,32 @@
 <?php
+require_once 'functions.php';
+require_once '../db/db_query.php';
+
+$db = getDbConnection();
+$newParagraph = $_POST['edit_form'];
```
Comment (JKapella) [inline php/admin_page.php]: This should be protected by an isset function!

File: php/admin_page.php [hunk 14]
```diff
@@ -1,4 +1,32 @@
...
+$addSubmit = $_POST['add_form'];
```
Comment (JKapella) [inline php/admin_page.php]: This should be protected by an isset function!

File: php/admin_page.php [hunk 15]
```diff
@@ -1,4 +1,32 @@
...
+$paragraphs = addAboutMe($db);
+$editDropdownResults = editParagraphDropdown($paragraphs);
+
```
Comment (JKapella) [inline php/admin_page.php:8]: don't need this blank line

File: php/admin_page.php [hunk 16]
```diff
@@ -1,4 +1,32 @@
...
+    $editChoice = $_POST['editSelect'];
+    $getEdit = getEdit($db, $editChoice);
+    $pasteEdit = pasteEdit($getEdit);
+}
+
+
+if (isset($_POST['submit'])){
+    $submitChoice = $_POST['editId'];
+    editParagraph($db, $submitChoice, $newParagraph);
+    header('Location: admin_page.php');
```
Comment (JKapella) [inline php/admin_page.php]: Do you need these extra redirects? It seems a bit wierd to refresh the page and then refresh the page again?

File: php/admin_page.php [hunk 17]
```diff
@@ -1,4 +1,32 @@
...
+}
+
+if (isset($_POST['delete'])){
+    $deleteChoice = $_POST['deleteSelect'];
+    deleteParagraph($db, $deleteChoice);
+    header('Location: admin_page.php');
```
Comment (JKapella) [inline php/admin_page.php]: Likewise, do you really need this redirect? Seems a bit redundant?

File: php/admin_page.php [hunk 18]
```diff
@@ -10,50 +38,61 @@
...
+            <form method="post" class="formText" action="admin_page.php">
                 <h4>Choose which paragraph to edit</h4>
                 <select name="editSelect">
-                    <option>paragraph 1</option>
-                    <option>paragraph 2</option>
-                    <option>paragraph 3</option>
+                    <option selected="selected" value="choose">Choose Paragraph</option>
+                    <?php if(isset($editDropdownResults)){
+                        echo $editDropdownResults;
+                    } ?>
```
Comment (JKapella) [inline php/admin_page.php]: These issets should be set at the top in the rest of the php I think. Not down in the html?

File: php/admin_page.php [hunk 19]
```diff
@@ -10,50 +38,61 @@
...
                 </select>
+                <div class="edit">
+                    <input type="submit" name="edit" value="Edit">
+                </div>
+            </form>
+            <form method="post" class="formText editParaRight" action="admin_page.php">
+                <textarea class="edit_submit" name="edit_form" required><?php
+                    if(isset($pasteEdit)){
```
Comment (JKapella) [inline php/admin_page.php:83]: Likewise here, i'd add this stuff in at the top?

File: php/admin_page.php [hunk 20]
```diff
@@ -10,50 +38,61 @@
...
+                        echo $pasteEdit;
+                    } ?></textarea>
+                <?php if (isset($editChoice)){
```
Comment (JKapella) [inline php/admin_page.php:86]: Likewise here!

File: php/admin_page.php [hunk 21]
```diff
@@ -10,50 +38,61 @@
...
+                    echo '<input type="hidden" name="editId" value=' . $editChoice . ' />';
```
Comment (JKapella) [inline php/admin_page.php]: I feel like this stuff should all be done up in the php at the top?

File: php/admin_page.php [hunk 22]
```diff
@@ -10,50 +38,61 @@
...
+                }
+                ?>
                 <div class="submit_buttons">
-                    <input type="submit" name="textEdit" value="Edit Text">
-                    <input type="submit" name="textSubmit" value="Submit New Text">
+                    <?php $submitButton = submitButton(); ?>
```
Comment (JKapella) [inline php/admin_page.php]: And dis

File: php/admin_page.php [hunk 23]
```diff
@@ -10,50 +38,61 @@
...
             <p>Delete paragraph</p>
-            <form>
+            <form method="post" action="admin_page.php">
                 <h4>Choose which paragraph to delete</h4>
                 <select name="deleteSelect">
-                    <option>paragraph 1</option>
-                    <option>paragraph 2</option>
-                    <option>paragraph 3</option>
+                    <option selected="selected" value="choose">Choose Paragraph</option>
+                    <?php if(isset($editDropdownResults)){
```
Comment (JKapella) [inline php/admin_page.php]: This one too!

File: php/functions.php [hunk 24]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
  * @param $addSubmit string inserts the string in the text area to the database and pairing it with an `id` and a `deleted` number
  */
 function addParagraphToDb (PDO $db, string $addSubmit) : void{
-    $query = $db->prepare("INSERT INTO `about_me` (`paragraph`) VALUES (:newParagraph);");
-    $query->bindParam(':newParagraph', $addSubmit);
-    $query->execute();
+    if (empty(trim($addSubmit))){
+        header('Location: admin_page.php');
```
Comment (JKapella) [inline php/functions.php]: Don't think you should have headers here in the functions redirecting you, particularly as there's nothing in the function name of docblock to suggest it redirects you

File: php/functions.php [hunk 25]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
+    } else if (strlen($addSubmit) < 1000) {
+        $query = $db->prepare("INSERT INTO `about_me` (`paragraph`) VALUES (:newParagraph);");
+        $query->bindParam(':newParagraph', $addSubmit);
+        $query->execute();
+    } else {
+        header('Location: admin_page.php');
```
Comment (JKapella) [inline php/functions.php]: Again here, I wouldn't do these, i'd just return false or something else like that instead

File: php/functions.php [hunk 26]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
+        $query->bindParam(':newParagraph', $newParagraph);
+        $query->bindParam(':editChoice', $submitChoice);
+        return $query->execute();
+    } else {
+        header('Location: admin_page.php');
+        }
+    }
 
 
+function submitButton(){
```
Comment (JKapella) [inline php/functions.php]: No docblock

File: php/functions.php [hunk 27]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
+/**
+ * this function enables the html page to call and put the paragraph that is chosen from the dropdown and into the textarea for edit
+ *
+ * @param array $getEdit is the variable that retrieves the output of the getEdit function, which is the selected paragraph in the dropdown
+ *
+ * @return string returns a variable that can be assigned in the admin page and can be put into the textarea, the textarea will be populated by the chosen paragraph
+ */
+function pasteEdit(array $getEdit) : string {
+    $editPopulate = '';
+    if (is_string($getEdit['paragraph']) &&array_key_exists('paragraph', $getEdit)) {
```
Comment (JKapella) [inline php/functions.php]: Need a space between the && and function name

File: php/functions.php [hunk 28]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
 
+//function addParagraphToDb (PDO $db, string $addSubmit) : void{
+//    if (strlen($addSubmit) < 1000) {
+//        $query = $db->prepare("INSERT INTO `about_me` (`paragraph`) VALUES (:newParagraph);");
+//        $query->bindParam(':newParagraph', $addSubmit);
+//        $query->execute();
+//    } else {
+//        header('Location: admin_page.php');
+//    }
+//}
```
Comment (JKapella) [inline php/functions.php]: Remove commented out code

File: php/functions.php [hunk 29]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
+/**
+ * this function retrieves the paragraph that is chosen in the dropdown list
+ *
+ * @param $db PDO calls the database from the db_query because this function requires it
+ *
+ * @param $editChoice string outputs a number of the id that will be used by the WHERE in MySQL
+ *
+ * @return mixed retrieves one row of the result by the SELECT in MySQL
+ */
+function getEdit(PDO $db, string $editChoice){
```
Comment (JKapella) [inline php/functions.php]: Missing return type on function

File: php/functions.php [hunk 30]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
 }
 
+/**
+ * this function retrieves the paragraph that is chosen in the dropdown list
+ *
+ * @param $db PDO calls the database from the db_query because this function requires it
+ *
+ * @param $editChoice string outputs a number of the id that will be used by the WHERE in MySQL
+ *
+ * @return mixed retrieves one row of the result by the SELECT in MySQL
```
Comment (JKapella) [inline php/functions.php]: Need to make a consistent return value

File: php/functions.php [hunk 31]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
+ */
+function getEdit(PDO $db, string $editChoice){
+    if ($editChoice == "choose"){
+        header('Location: admin_page.php');
```
Comment (JKapella) [inline php/functions.php]: Don't think you should have a redirect in here!

File: php/functions.php [hunk 32]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
+ *
+ * @param $editChoice string outputs a number of the id that will be used by the WHERE in MySQL
+ *
+ * @param $newParagraph string is a variable that gets the text that was submitted in the 'edit paragraph' text area
+ *
+ * @return bool the function will only be called when the user press 'submit' and if true, will update the selected paragraph with the new paragraph
+ */
+function editParagraph(PDO $db, string $submitChoice, string $newParagraph) : bool {
+    if (empty(trim($newParagraph))){
+        header('Location: admin_page.php');
```
Comment (JKapella) [inline php/functions.php]: Redirect shouldn't be here in your function methinks

File: php/functions.php [hunk 33]
```diff
@@ -47,30 +42,126 @@ function addParagraphs(array $paragraphs) :string {
...
+    } else if (strlen($newParagraph) < 1000) {
+        $query = $db->prepare("UPDATE `about_me` SET `paragraph`= :newParagraph WHERE `id`= :editChoice;");
+        $query->bindParam(':newParagraph', $newParagraph);
+        $query->bindParam(':editChoice', $submitChoice);
+        return $query->execute();
+    } else {
+        header('Location: admin_page.php');
+        }
+    }
```
Comment (JKapella) [inline php/functions.php]: Formatting on these end curly braces needs fixing

File: php/tests.php [hunk 34]
```diff
@@ -41,4 +41,58 @@ public function testAddParagraphsMalformed(){
         addParagraphs($input);
     }
 
+
```
Comment (JKapella) [inline php/tests.php]: Please add the tests to a tests folder and rename this!

**Review by JKapella — COMMENTED** [review]
Reviewed - some things to pick up on!

File: php/admin_page.php (hunk 16 above)
Comment (rrrichard) [reply php/admin_page.php]: charlie said its alright to redirect

File: php/admin_page.php (hunk 17 above)
Comment (rrrichard) [reply php/admin_page.php]: deleted it

File: php/admin_page.php (hunk 18 above)
Comment (rrrichard) [reply php/admin_page.php]: apparently if it relies on $_POSTs I need to do an isset on top and one when its called

File: php/admin_page.php (hunk 19 above)
Comment (rrrichard) [reply php/admin_page.php:83]: apparently if it relies on $_POSTs I need to do an isset on top and one when its called

File: php/admin_page.php (hunk 18 above)
Comment (rrrichard) [reply php/admin_page.php]: wooops my bad, this one doesnt need an isset because its always loaded with the page

File: php/admin_page.php (hunk 21 above)
Comment (rrrichard) [reply php/admin_page.php]: done, made into a variable at the top

File: php/admin_page.php (hunk 22 above)
Comment (rrrichard) [reply php/admin_page.php]: yeap done it in $_POST['edit']

File: php/admin_page.php (hunk 23 above)
Comment (rrrichard) [reply php/admin_page.php]: no need for this one as dropdown is always present when page loads, removed the isset

File: php/functions.php (hunk 24 above)
Comment (rrrichard) [reply php/functions.php]: removed this from function, added it into the admin_add in an if statement

File: php/functions.php (hunk 25 above)
Comment (rrrichard) [reply php/functions.php]: same as the one up

File: php/functions.php (hunk 26 above)
Comment (rrrichard) [reply php/functions.php]: yeap added

File: php/functions.php (hunk 27 above)
Comment (rrrichard) [reply php/functions.php]: done

File: php/functions.php (hunk 28 above)
Comment (rrrichard) [reply php/functions.php]: got it dude

File: php/functions.php (hunk 29 above)
Comment (rrrichard) [reply php/functions.php]: yeap added : array

File: php/functions.php (hunk 30 above)
Comment (rrrichard) [reply php/functions.php]: changed to retrieves an array

File: php/functions.php (hunk 31 above)
Comment (rrrichard) [reply php/functions.php]: changed it like all the other functions with redirect, the if statement is in admin_page

File: php/functions.php (hunk 32 above)
Comment (rrrichard) [reply php/functions.php]: nowRichard thinks so too!

File: php/functions.php (hunk 33 above)
Comment (rrrichard) [reply php/functions.php]: deleted those because of your previous comments

File: php/tests.php (hunk 34 above)
Comment (rrrichard) [reply php/tests.php]: wohooooo final comment! done it

**Review by JKapella — APPROVED** [review]
YES

**Pull request merged**